fvm/state/state_replay_cache.go

```diff
@@ -0,0 +1,87 @@
+package state
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StateReplayCache caches value read from state together with all register touches triggered
+// when reading that value from state. These register touches will be re/played on the target
+// view in every GetOrRetrieveValue value call.
+type StateReplayCache[T any] struct {
+	cachedValue T
+	cachedState *State
+}
+
+func NewStateReplayCache[T any]() *StateReplayCache[T] {
+	return &StateReplayCache[T]{}
+}
+
+// GetOrRetrieveValue returns a state value. When value is not cached, `retriver` closure will be
+// called to retrieve the value and the state delta will be captured and cached too; when value is
+// cached the cached one will be returned, and the cached state delta will be replayed on `view`.
+func (stateReplayCache *StateReplayCache[T]) GetOrRetrieveValue(
+	view View,
```

Comment:
maybe pass in the transactionState rather than the view.  this enables us to compose nested transactions.

Reply:
i was thinking about that, but in the case of meter settings, in `fvm.Run` when a transactionState is created, meter settings is needed: https://github.com/onflow/flow-go/blob/master/fvm/fvm.go#L122, so if we pass in transactionState, we will need to create a separate transactionState and then manually Pause\Assume that state. And since state here is mostly for a View, passing a View can make that use cases slightly more elegant.

Reply:
o, i take my comment back.  in general, we need to pass in the transaction state rather than the view because we need update the transaction state's meter / spock hash, which are not part of the view